packages/store/ts/config/v2/table.ts

```diff
@@ -0,0 +1,145 @@
+import { ErrorMessage, evaluate } from "@arktype/util";
+import { SchemaInput, resolveSchema } from "./schema";
+import { AbiTypeScope, getStaticAbiTypeKeys } from "./scope";
+
+export type NoStaticKeyFieldError =
+  ErrorMessage<"Provide a `key` field with static ABI type or a full config with explicit primaryKey override.">;
+
+export type ValidKeys<schema extends SchemaInput<scope>, scope extends AbiTypeScope> = [
+  getStaticAbiTypeKeys<schema, scope>,
+  ...getStaticAbiTypeKeys<schema, scope>[]
+];
+
+export type TableInput<
+  schema extends SchemaInput<scope> = SchemaInput,
+  scope extends AbiTypeScope = AbiTypeScope,
+  primaryKey extends ValidKeys<schema, scope> = ValidKeys<schema, scope>
+> = TableFullInput<schema, scope, primaryKey> | TableShorthandInput<scope>;
+
+export type TableShorthandInput<scope extends AbiTypeScope = AbiTypeScope> =
+  | SchemaInput<scope>
+  | keyof scope["validTypes"];
```

Comment:
chatted last night about potentially removing this

I am fine keeping this here for now but should make sure we consider what we wanna expose/ship with because it'll be harder to do once this is cemented into the API